### PR TITLE
Convert tests to pytest (solvers and internals)

### PR DESCRIPTION
**Pull request by jakelishman**
These are the files that Eric (@Ericgig) was checking in #1181, but the commits are tidied up and rebased onto `master`.  I'll copy over the review comments in that PR that haven't been actioned yet.  The potential merge conflict with `test_mcsolve.py` is fixed.

This is based on #1249, and that one needs to be merged first.

**Changelog**
Major test refactor into a pytest style.

### CODE REVIEW

File: qutip/tests/test_interpolate.py [hunk 1]
```diff
@@ -30,325 +30,123 @@
...
+# the test, and `type` is a string in ["spline", "function", "string"] which
+# tells us how to interpret the type of `other`.  If `type` is "function" or
+# "string", then nothing is done to `other` and it is used as-is.  If `type` is
+# "spline", then `other` is given as a vectorised function which we turn into a
+# qutip.Cubic_Spline using appropriate `x` and `y` values.
+_case_real_spline = ('0.25*sin(t)', lambda t: 0.25*np.sin(t), 'spline')
+_case_real_string = ('0.25*cos(t)', '0.25*cos(t)', 'string')
+_case_real_function = ('0.25*sin(t)', lambda t, *_: 0.25*np.sin(t), 'function')
+_case_complex_spline = ('0.1j*sin(t)', lambda t: 0.1j*np.sin(t), 'spline')
+_case_complex_string = ('-0.1j*sin(t)', '-0.1j*sin(t)', 'string')
```
Comment (jakelishman) [inline qutip/tests/test_interpolate.py:86]: Comment from #1181 by @Ericgig:
>The reference could be converted to a function to save time.

My response:
>Ah, so the original reason that the base was a string was to accommodate brmesolve, which can't accept the function forms. I suspect they were originally just copy+paste code, so they were all the same.
>
>Are you planning a complete overhaul of brmesolve? If so, this switch-over could wait until then. If not, I'll think rather more about how to handle these cases so it doesn't turn into a huge mess.

**Comment by Ericgig** [comment]
- [x] test_mcsolve.py
- [x] test_lattice.py
- [x] test_interpolate.py
- [ ] test_fastsparse.py
- [ ] test_expect.py
- [ ] test_eigenstates.py
- [ ] test_dimensions.py
- [ ] test_cy_structs.py

File: qutip/tests/test_expect.py [hunk 2]
```diff
@@ -31,151 +31,149 @@
...
+
+@pytest.mark.repeat(20)
+@pytest.mark.parametrize("hermitian", [False, True], ids=['complex', 'real'])
+def test_equivalent_to_matrix_element(hermitian):
+    dimension = 20
+    state = qutip.rand_ket(dimension, 0.3)
+    op = qutip.rand_herm(dimension, 0.2)
+    if not hermitian:
+        op = op + 1j*qutip.rand_herm(dimension, 0.1)
+    matrix = (state.dag() * op * state).data[0, 0]
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply qutip/tests/test_expect.py:157]: Yeah, it was meant to stand for "matrix element".  I'll just change it to "expected".

File: qutip/tests/test_expect.py [hunk 3]
```diff
@@ -31,151 +31,149 @@
...
+    pytest.param(qutip.sesolve, id="sesolve"),
+    pytest.param(functools.partial(qutip.mesolve, c_ops=[0.001 * _sz]),
+                 id="mesolve"),
+])
+def test_compatibility_with_solver(solve):
+    e_ops = [getattr(qutip, 'sigma'+x)() for x in 'xyzmp']
+    h = qutip.sigmax()
+    state = qutip.basis(2, 0)
+    times = np.linspace(0, 10, 101)
+    direct = solve(h, state, times, e_ops=e_ops).expect
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply qutip/tests/test_expect.py:171]: Done, good point, thanks!

File: qutip/tests/test_interpolate.py [hunk 4]
```diff
@@ -30,325 +30,122 @@
...
+# reference), `other` is what will become the time-dependence for this part of
+# the test, and `type` is a string in ["spline", "function", "string"] which
+# tells us how to interpret the type of `other`.  If `type` is "function" or
+# "string", then nothing is done to `other` and it is used as-is.  If `type` is
+# "spline", then `other` is given as a vectorised function which we turn into a
+# qutip.Cubic_Spline using appropriate `x` and `y` values.
+_case_real_spline = ('0.25*sin(t)', lambda t: 0.25*np.sin(t), 'spline')
+_case_real_string = ('0.25*cos(t)', '0.25*cos(t)', 'string')
+_case_complex_spline = ('0.1j*sin(t)', lambda t: 0.1j*np.sin(t), 'spline')
+_case_complex_string = ('-0.1j*sin(t)', '-0.1j*sin(t)', 'string')
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply qutip/tests/test_interpolate.py]: Converted this to a nicer form which is hopefully a bit easier to read, and as a side-effect supports pickling!

File: qutip/tests/test_interpolate.py [hunk 5]
```diff
@@ -30,325 +30,122 @@
...
+    H_test = H_expected.copy()
+    operator = a**2 + a.dag()**2
+    for string, other, type_ in coefficients:
+        if type_ == 'spline':
+            other = qutip.Cubic_Spline(times[0], times[-1], other(times))
+        H_expected.append([operator, string])
+        H_test.append([operator, other])
+    expected = solver(H_expected, state, times, e_ops=e_ops).expect[0]
+    test = solver(H_test, state, times, e_ops=e_ops).expect[0]
+    np.testing.assert_allclose(test, expected, atol=1e-4)
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply qutip/tests/test_interpolate.py:151]: Done at the slight cost of a weirder test signature.

**Comment by jakelishman** [comment]
@Ericgig: all fixed up and all your changes made.  Still needs #1249 to be reviewed before this can be merged.

**Comment by Ericgig** [comment]
@jakelishman merge or squash this one? Seems a few of the commits were also in #1249.

**Comment by jakelishman** [comment]
@Ericgig: the commits are the same ones you just merged in (as in the same SHA-1 - they're completely identical git objects), so they shouldn't appear if you do a regular merge - the fact they still appear here _should_ just be that GitHub doesn't update its UI when other PRs are merged.

I suggest regular merges because I've tidied up the commit history in these PRs, so it's easier to search back through the commit history.

**Comment by jakelishman** [comment]
I'm glad I was right and I didn't just make a mess!

**Pull request merged**